lib/membrane/timestamp_queue.ex

```diff
@@ -275,112 +290,104 @@ defmodule Membrane.TimestampQueue do
-        {:infinity, pad_ref} -> do_pop(timestamp_queue, pad_ref)
-        _other -> {:none, timestamp_queue}
-      end
+      _other -> pop_following_items(timestamp_queue, pad_ref, actions_acc, items_acc)
     end
   end
 
-  @spec do_pop(t(), Pad.ref()) :: {popped_value() | :none, t()}
-  defp do_pop(timestamp_queue, pad_ref) do
-    pad_queue = Map.get(timestamp_queue.pad_queues, pad_ref)
+  defp pop_following_items(%__MODULE__{} = timestamp_queue, pad_ref, actions_acc, items_acc) do
+    pad_queue = timestamp_queue.pad_queues |> Map.get(pad_ref)
 
     case Qex.pop(pad_queue.qex) do
-      {{:value, {:buffer, buffer}}, qex} ->
-        buffer_time = buffer_time(buffer, pad_queue)
-
-        case pad_queue do
-          %{update_heap_on_buffer?: true} ->
-            timestamp_queue
-            |> Map.update!(:pads_heap, &(&1 |> Heap.pop() |> Heap.push({-buffer_time, pad_ref})))
-            |> put_in([:pad_queues, pad_ref, :update_heap_on_buffer?], false)
-            |> do_pop()
-
-          %{buffers_number: 1, end_of_stream?: false} ->
-            # last buffer on pad queue without end of stream
-            {:none, timestamp_queue}
-
-          pad_queue ->
-            buffer_size = timestamp_queue.metric.buffers_size([buffer])
-
-            pad_queue = %{
-              pad_queue
-              | qex: qex,
-                buffers_size: pad_queue.buffers_size - buffer_size,
-                buffers_number: pad_queue.buffers_number - 1,
-                update_heap_on_buffer?: true
-            }
-
-            timestamp_queue =
-              timestamp_queue
-              |> Map.put(:current_queue_time, buffer_time)
-              |> put_in([:pad_queues, pad_ref], pad_queue)
-
-            {{pad_ref, {:buffer, buffer}}, timestamp_queue}
-        end
+      {{:value, {:buffer, buffer}}, _qex} ->
+        # update heap, ignore buffer, recursion to pop_batch
+        new_priority = -buffer_time(buffer, pad_queue)
+        heap_item = {new_priority, pad_ref}
+
+        timestamp_queue
+        |> Map.update!(:pads_heap, &(&1 |> Heap.pop() |> Heap.push(heap_item)))
+        |> do_pop_batch(actions_acc, items_acc)
 
       {{:value, item}, qex} ->
-        timestamp_queue =
-          timestamp_queue
-          |> put_in([:pad_queues, pad_ref, :qex], qex)
+        # add to acc, recursion self
+        timestamp_queue = put_in(timestamp_queue, [:pad_queues, pad_ref, :qex], qex)
+        items_acc = [{pad_ref, item}] ++ items_acc
 
-        {{pad_ref, item}, timestamp_queue}
+        pop_following_items(timestamp_queue, pad_ref, actions_acc, items_acc)
 
-      {:empty, _qex} ->
+      {:empty, _empty_qex} when timestamp_queue.awaiting_pads == [pad_ref] ->
+        # pop heap, recursion to pop_batch
         timestamp_queue
-        |> Map.update!(:pad_queues, &Map.delete(&1, pad_ref))
         |> Map.update!(:pads_heap, &Heap.pop/1)
-        |> do_pop()
-    end
-  end
+        |> do_pop_batch(actions_acc, items_acc)
```

Comment:
This makes the flow harder to follow, can we return the accumulators here to `do_pop_batch` instead, so that each function only calls itself recursively?